### PR TITLE
Fix validation of certificates with CN=*

### PR DESCRIPTION
**Pull request by sp-andwei**
The bug was discovered and detailed by Joshua Rogers at
https://megamansec.github.io/Squid-Security-Audit/
where it was filed as "Buffer UnderRead in SSL CN Parsing".

### CODE REVIEW

**Review by rousskov — APPROVED** [review]
@sp-andwei, thank you for posting this fix!

I adjusted PR title/description and fixed Uri.cc [whitespace](https://github.com/squid-cache/squid/actions/runs/6552791874/job/17796847885?pr=1523) to get this PR into mergeable state. Please adjust further as needed.

@MegaManSec, FYI: you are mentioned in the PR description (i.e. the future commit message). I could not include the full vulnerability-specific [URL](https://megamansec.github.io/Squid-Security-Audit/ssl-bufferunderread.html) because it is longer than our 72 character limit, but I think referring to the top page and citing the vulnerability report title is enough.

**Comment by rousskov** [comment]
OK to test

**Comment by sp-andwei** [comment]
Thanks @rousskov. Just noticed I messed up the whitespace and wanted to fix but you were faster

**Review by yadij — APPROVED** [review]

**Review by kinkie — APPROVED** [review]

**Comment by kinkie** [comment]
@rousskov , given this is unanimously approved and a blocker for 6.X release, are we ok to hand-merge to expedite?

**Comment by squidadm** [comment]
queued for backport to v6

**Comment by yadij** [comment]
@rousskov as you know Anubis does not follow the voting rules the same way we would. FYI, this is the behaviour I have observed __actually__ occurring:

**Rule 1)** `One negative vote by a core developer blocks the merge until resolved.`

Enforced by github policy, not Anubis.

**Rule 2)** ` Two positive votes from core developers accept the submission (with a high priority for merging).`

Anubis interprets `a high priority for merging` as switching to 48hr delayed "Fast Track"

**Rule 3)** `Any three positive votes accept the submission.`

Anubis previously has been to only "Fast-Track" these same as 2-vote PRs.

... thus the question @kinkie posted (on behalf of both us maintainers), assuming we would have to wait 48hrs without explicit intervention. If this 14min turnaround for merge on 3 core votes is going to continue this is good to know.

**Rule 4)** `Submissions older than 10 days without negative votes are accepted.`

Anubis waits for manual `M-cleared-for-merge` label before attempting to merge.

FWIW,  PR #878 has passed this criteria. But none of the automation has merged it.

**Comment by rousskov** [comment]
> @rousskov as you know Anubis does not follow the voting rules the same way we would.

I do not know that. In fact, I believe that Anubis does follow the voting rules pretty much "the same way we would" do that in a GitHub environment, in any meaningful sense of that phrase.

> **Rule 1)** `One negative vote by a core developer blocks the merge until resolved.`

> Enforced by github policy, not Anubis.

Enforced by Anubis (and probably by GitHub policy as well, but that detail is not relevant to this discussion).

 
> **Rule 2)** ` Two positive votes from core developers accept the submission (with a high priority for merging).`
 
> Anubis interprets `a high priority for merging` as switching to 48hr delayed "Fast Track"

Yes, no deviation from rules here. Needless to say, 48 hours is "high priority" compared to the regular 10-day timeout.

 
> **Rule 3)** `Any three positive votes accept the submission.`
 
> Anubis previously has been to only "Fast-Track" these same as 2-vote PRs.

Anubis is configured to treat PRs with three core developer votes as approved immediately. Again, no deviation from rules here as far as this discussion is concerned.

I am not sure why you mentioned what Anubis did "previously". The change to treat unanimous core developer votes (currently the three of us, for all practical purposes) specially has been discussed (by all three of us), agreed upon, and deployed long time ago.

> If this 14min turnaround for merge on 3 core votes is going to continue this is good to know.

The time a PR can be merged after it clears the voting hurdle depends on several factors. Bugs notwithstanding, all those factors are beyond Anubis control. For example, in many cases, CI tests will need to be executed and will take much longer to complete than 14 minutes.


> **Rule 4)** `Submissions older than 10 days without negative votes are accepted.`

Yes, Anubis enforces this rule for PRs submitted or explicitly approved by a core developer.

Due to GitHub realities, Anubis does not implement (or violate) this rule for PRs that were not approved by core developers: We have to rely on a core developer (or equivalent) to clear PRs for merging. That core developer should approve the PR, making this remaining scope of this rule effectively inapplicable to Anubis.

If you insist, we can enhance Anubis to implement Rule 4 for _all_ developers instead of requiring one of us to explicitly approve PRs by non-core developers. I strongly do _not_ recommend doing this (I can detail why if needed), but we _can_ do this if you insist.
 
> Anubis waits for manual `M-cleared-for-merge` label before attempting to merge.

While Anubis is indeed configured to require that label before merging, the label does not affect (or violate) voting rules. It is a separate control layer, just like our CI tests and GitHub git conflicts checks are. Adding that label to a PR that does not have any votes will not allow Anubis to merge that PR. This label is very similar, in its effect, to how some core developer still had to commit the "approved" patch in the old days, before GitHub/Anubis. These days, we label PRs. We can turn this Anubis feature off, but I do not recommend that.

> FWIW, PR #878 has passed this criteria. But none of the automation has merged it.

Exactly. That PR is "accepted" (per our voting rules). Now we need a core developer to validate that acceptance (by approving the PR, with an "approving without review by Project Voting rule 4" disclaimer if necessary) and clear it for merging (by labeling the PR). FWIW, this is similar to when we were merging squid-dev patches by hand -- a core developer also had to judge acceptance and do something for the "accepted" patch to be actually merged...

**Pull request closed without merging**
